core/consensus/grandpa/grandpa_context.hpp

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_CONSENSUS_GRANDPACONTEXT
+#define KAGOME_CONSENSUS_GRANDPACONTEXT
+
+#include "consensus/grandpa/structs.hpp"
+#include "network/types/grandpa_message.hpp"
+
+#include <libp2p/peer/peer_id.hpp>
+
+namespace kagome::consensus::grandpa {
+
+  class GrandpaContext final {
+   public:
+    // RAII holder of context on the stack
+    class Guard {
+      // Avoid creation of guard in the heap
+      void *operator new(size_t) = delete;
+      void *operator new(size_t, void *) = delete;
+      void *operator new[](size_t) = delete;
+      void *operator new[](size_t, void *) = delete;
+
+     public:
+      Guard() {
+        // Create context in ctor
+        create();
+      }
+
+      ~Guard() {
+        // Release context in the out of scope
+        release();
+      }
+    };
+
+    // Payload
+    std::optional<const libp2p::peer::PeerId> peer_id{};
+    std::optional<const network::VoteMessage> vote{};
+    std::optional<const network::CatchUpResponse> catch_up_response{};
+    std::optional<const network::FullCommitMessage> commit{};
+    std::set<primitives::BlockInfo, std::greater<primitives::BlockInfo>>
+        missing_blocks{};
+
+    static void set(std::shared_ptr<GrandpaContext> context) {
+      auto opt = instance();
+      BOOST_ASSERT(not opt.has_value());
+      opt.emplace(std::move(context));
+    }
+
+    static std::optional<std::shared_ptr<GrandpaContext>> get() {
+      auto opt = instance();
+      return opt;
+    }
+
+   private:
+    static std::optional<std::shared_ptr<GrandpaContext>> &instance() {
+      static thread_local std::optional<std::shared_ptr<GrandpaContext>> inst;
+      return inst;
+    }
+
+    static void create() {
+      auto &opt = instance();
+      BOOST_ASSERT(not opt.has_value());
```

Comment:
I am a little afraid that if for some reasons there will be 2 contexts created on the same threard and our program is running in release mode, we might get some unexpected issues, which will be hard to fix. Maybe we should print some meaningfull message or terminate program in that case?

Reply:
Logging is inappropriate here, because it is static method. I think throwing of exception is acceptable way for mentioned problem. I'll add it.